src/program.cpp

```diff
@@ -111,342 +110,115 @@ program& program::operator=(program p)
-}
-
-instruction_ref program::add_return(std::vector<instruction_ref> args)
-{
-    assert(std::all_of(
-               args.begin(), args.end(), [&](instruction_ref x) { return has_instruction(x); }) &&
-           "Argument is not an exisiting instruction");
-    impl->instructions.push_back({builtin::returns{}, {}, args});
-    auto result = std::prev(impl->instructions.end());
-    instruction::backreference(result);
-    assert(result->valid(begin()));
-
-    return result;
-}
-
-shape program::get_parameter_shape(std::string name) const
+std::unordered_map<std::string, shape> program::get_parameter_shapes() const
 {
-    auto ins = std::find_if(
-        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-            if(x.name() == "@param")
-            {
-                return any_cast<builtin::param>(x.get_operator()).parameter == name;
-            }
-            else
-            {
-                return false;
-            }
-        });
-    if(ins != this->end())
-        return ins->get_shape();
-    else
-        return {};
+    assert(contains(impl->modules, "main"));
+    return impl->modules["main"].get_parameter_shapes();
 }
 
-std::vector<std::string> program::get_parameter_names() const
+bool program::has_instruction(instruction_ref ins) const
 {
-    std::vector<std::string> result = impl->input_names;
-    std::unordered_set<std::string> params;
-    for(auto&& ins : impl->instructions)
-    {
-        if(ins.name() == "@param")
-        {
-            auto&& name = any_cast<builtin::param>(ins.get_operator()).parameter;
-            params.insert(name);
-        }
-    }
-    erase_if(result, [&](auto&& name) { return params.count(name) == 0; });
-    return result;
+    assert(contains(impl->modules, "main"));
+    return impl->modules["main"].has_instruction(ins);
 }
 
-instruction_ref program::get_parameter(std::string name) const
-{
-    auto ins = std::find_if(
-        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-            if(x.name() == "@param")
-            {
-                return any_cast<builtin::param>(x.get_operator()).parameter == name;
-            }
-            else
-            {
-                return false;
-            }
-        });
-    if(ins != this->end())
-        return ins;
-    else
-        return this->end();
-}
+std::size_t program::size() const { return impl->modules.size(); }
 
-std::unordered_map<std::string, shape> program::get_parameter_shapes() const
+instruction_ref program::begin() const
 {
-    std::unordered_map<std::string, shape> result;
-    for(auto&& ins : impl->instructions)
-    {
-        if(ins.name() == "@param")
-        {
-            auto&& name  = any_cast<builtin::param>(ins.get_operator()).parameter;
-            result[name] = ins.get_shape();
-        }
-    }
-    return result;
+    assert(contains(impl->modules, "main"));
+    return impl->modules["main"].begin();
```

Comment:
Could you call `this->get_main_module()` instead of `impl->modules["main"]`? This can help get rid of the repeated asserts.

Reply:
fixed